src/ocean/util/prometheus/collector/CollectorRegistry.d

```diff
@@ -0,0 +1,209 @@
+    /// The collector to use for stat collection. This acts as the actual
+    /// parameter for the delegates in `collector_callbacks`.
+    private Collector collector;
+
+    /***************************************************************************
+
+        Constructor, that accepts an array of stat collection callbacks.
+
+        Params:
+            callbacks : An array of delegates to be called for stat collection.
+
+    ***************************************************************************/
+
+    public this ( CollectionDg[] callbacks )
+    {
+        this.collector = new Collector();
+        this.collector_callbacks = callbacks;
+    }
+
+    /***************************************************************************
+
+        Appends a stat collector callback to the list of existing ones.
+        May result in a re-allocation of the entire buffer storing the
+        callbacks.
+
+        Params:
+            collector : The stat collector to append.
+
+    ***************************************************************************/
+
+    public void addCollector ( CollectionDg collector )
+    {
+        this.collector_callbacks ~= collector;
+    }
+
+    /***************************************************************************
+
+        Collects stats by calling all delegates specified for stat collection
+        and returns the accumulated result.
+
+        Returns:
+            The stats accumulated since last time this method was called.
+
+    ***************************************************************************/
+
+    public cstring collect ( )
+    {
+        this.collector.reset();
+
+        try
+        {
+            foreach (ref callback; this.collector_callbacks)
+            {
+                callback(this.collector);
+            }
+        }
+        catch (Exception ex)
+        {
+            throw ex;
+        }
+
+        return this.collector.getCollection();
+    }
+}
+
+version (UnitTest)
+{
+    import ocean.core.Test;
+    import Traits = ocean.core.Traits;
+    import ocean.transition;
+    import ocean.util.prometheus.collector.Collector;
+
+    private struct Statistics
+    {
+        ulong up_time_s;
+        size_t count;
+        float ratio;
+        double fraction;
+        real very_real;
+    }
+
+    private struct Labels
+    {
+        hash_t id;
+        cstring job;
+        float perf;
+    }
+
+    Statistics test_stats;
+    Labels test_labels;
+
+    ///
+    private void setTestStats ( Statistics stats )
+    {
+        test_stats = stats;
+    }
+
+    ///
+    private void setTestLabels ( Labels labels )
+    {
+        test_labels = labels;
+    }
+
+    ///
+    public void collectDg1 ( Collector collector )
+    {
+        collector.collect!("id")(test_stats, 123.034);
+    }
+
+    ///
+    public void collectDg2 ( Collector collector )
+    {
+        collector.collect(test_stats, test_labels);
+    }
+}
+
+/// Test collection from a single delegate
```

Comment:
I removed usages of `toDg` from these tests (it looks like it was needed before because the calbacks were module-level functions). If it's alright, I have kept the tests commented with `///` to make them serve as examples as well.